### PR TITLE
Full parameter fine-tuning sampler sub-system 

### PR DESCRIPTION
**Pull request by droot**
**feat(k8s,vllm): introduce decoupled distributed reinforcement learning and vLLM time-slicing infrastructure**

### Overview
This PR introduces horizontally scalable, decoupled microservices architecture for running distributed Supervised Fine-Tuning (SFT) and Reinforcement Learning (RL) workloads on Kubernetes. By leveraging Kubernetes Dynamic Resource Allocation (DRA) exact allocation claims and cooperative VRAM yield, this design virtualizes GPU silicon to support high-density concurrent training and sampling experiments across shared hardware fleets.

### Key Architectural Highlights
- **Decoupled Workload Orchestration**: Separates policy gradient calculation (Trainers) from rollout generation (Samplers). The central Gateway service dynamically generates isolated Pod manifests per runtime session without requiring static worker deployments.
- **Dynamic Resource Allocation (DRA) Pinning**: Replaces standard exclusive GPU locks with exact-allocation device claims. Concurrent workloads co-schedule onto shared silicon up to configured concurrency limits.
- **Strict Role Segregation**: Introduces dedicated node affinity groups (`trainers` vs `samplers`) to guarantee physical hardware isolation between PyTorch AdamW optimizers and vLLM KV caches, preventing CUDA out-of-memory contention.
- **Cooperative Time-Slicing & NFS Weight Sync**: Virtualizes active GPU VRAM across multi-tenant inference engines. Sampler engines cooperatively sleep VRAM between batches and reload safetensor weights in-place in ~1.1 seconds when training updates synchronize over managed shared NFS storage.
- **Host Memory Oversubscription Tuning**: Optimizes worker CPU RAM requests (`16GiB`) to safely pack multiple concurrent distributed training jobs onto standard 48GiB host machines without scheduling deadlocks.

### End-to-End Verifications Achieved
- **10-Step Distributed RL Verification**: Successfully executed distributed reinforcement learning loops against live KRM clusters, verifying seamless loss drop and policy convergence to 1.00 mean reward (100% accuracy).
- **Dual Concurrent Workload Multiplexing**: Verified parallel execution of concurrent RL experiments sharing physical GPU silicon, confirming stable CRIU process swapping, cooperative inference yield, and zero-error cleanup upon session teardown.

### CODE REVIEW

File: k8s/deploy/distributed-fft-timeslice/04-gateway.yaml [hunk 1]
```diff
@@ -64,10 +64,14 @@ spec:
...
+          value: "/etc/open-rl/trainer/trainer-worker-pod.yaml"
+        - name: OPEN_RL_SAMPLER_POD_TEMPLATE
+          value: "/etc/open-rl/sampler/sampler-worker-pod.yaml"
         - name: OPEN_RL_WORKER_POD_TEMPLATE
-          value: "/etc/open-rl/trainer-worker-pod.yaml"
-        # Logical OpenRL snapshot group. It intentionally matches the
-        # timeslice.io/group label convention stamped onto trainer worker pods.
+          value: "/etc/open-rl/trainer/trainer-worker-pod.yaml"
+        - name: OPEN_RL_WORKER_IMAGE
+          value: "gcr.io/cdrollouts-sunilarora/open-rl-server:latest"
```
Comment (ShubyM) [inline k8s/deploy/distributed-fft-timeslice/04-gateway.yaml:74]: Should we add a kustomize flow to override this?

**Review by ShubyM — COMMENTED** [review]
Left one small comment otherwise looks good!
